Source/SampleBuffer.hpp

```diff
@@ -14,136 +14,172 @@
 #undef AUDIO_SATURATE_SAMPLES
 #endif //ARM_CORTEX
 
+#define MULTIPLIER_31B 0x80000000
+#define MULTIPLIER_30B 0x40000000
+#define MULTIPLIER_23B 0x00800000
 
 class SampleBuffer : public AudioBuffer {
 protected:
-  FloatArray left;
-  FloatArray right;
-  uint16_t size;
-  const float mul = 1/2147483648.0f;
+  FloatArray* buffers;
+  size_t blocksize;
+  size_t channels;
 public:
-  SampleBuffer(int blocksize){
-    left = FloatArray::create(blocksize);
-    right = FloatArray::create(blocksize);
+  SampleBuffer(size_t channels, size_t blocksize)
+    :channels(channels), blocksize(blocksize) {
+    buffers = new FloatArray[channels];
+    for(size_t i=0; i<channels; ++i)
+      buffers[i] = FloatArray::create(blocksize);
   }
   ~SampleBuffer(){
-    FloatArray::destroy(left);
-    FloatArray::destroy(right);
+    for(size_t i=0; i<channels; ++i)
+      FloatArray::destroy(buffers[i]);
+    delete[] buffers;
   }
-  void split32(int32_t* input, uint16_t blocksize){
-    size = blocksize;
-    for(int i=0; i<size; ++i){
-      left[i] = (int32_t)((*input++)<<8) * mul;
-      right[i] = (int32_t)((*input++)<<8) * mul;
+  virtual void split(int32_t* input) = 0;
```

Comment:
I don't think it's a good decision to use virtual functions here. Has template specialization been considered?

```
template<classname Data, uint8_t sample_bits>
class SampleBuffer {
     void split(Data* input);
};

inline void SampleBuffer<int32_t, 32>::split(int32_t* input) {
// ..
};
```

Something like this could be used to define implementations for different sample size. And really we could have a definition for <float, 32>, I think it would be preferable to add float streams processing to current PR to avoid refactoring code later.

Reply:
I want us to be able to compile patches that work across firmwares, regardless of channels and format. So templates don't work, unless there is still a common base class and then what is gained?
Up until now I have been avoiding virtual functions and making sure the conditional is outside of the block loop. But the cost of two virtual function calls per block is pretty small, for a big improvement in code maintainability.

Reply:
To clarify: the channel/format decision needs to be taken at runtime, not compile time, so templates don't optimise anything for us. Whether it is a better or worse choice of syntax is perhaps a matter of taste.

Reply:
Ok, I see what you mean.

But it would work if we ensure that firmware would be using the format with expected number of channels that patch requests, so basically it must match whatever would be passed as registerPatch arguments (the bigger number of IO channels). If we do it like that, then FW can switch sample format, and we only need to have a single template instance in compiled patch that would be able to handle different sample formats with the same number of channels.